test/test_modular_adaptor.cpp

```diff
@@ -0,0 +1,214 @@
+         BOOST_CHECK_EQUAL(a > b, (a_s % mod_s) > (b_s % mod_s));
+         BOOST_CHECK_EQUAL(a >= b, (a_s % mod_s) >= (b_s % mod_s));
+         BOOST_CHECK_EQUAL(a == b, (a_s % mod_s) == (b_s % mod_s));
+         BOOST_CHECK_EQUAL(a != b, (a_s % mod_s) != (b_s % mod_s));
+      }
+   }
+}
+
+template <typename Backend>
+void bitwise_operators()
+{
+   typedef number<modular_adaptor<Backend> > modular_number;
+   typedef modular_params<Backend>           params_number;
+   typedef number<Backend>                   standard_number;
+
+   for (size_t current_number_size = 0; current_number_size < 7; ++current_number_size)
+   {
+      for (size_t parity = 0; parity <= 1; ++parity)
+      {
+         std::string a_string = big_numbers[0].substr(0, numbers_size[current_number_size]);
+         std::string b_string = big_numbers[1].substr(0, numbers_size[current_number_size]);
+         std::string mod_string;
+         if (parity)
+            mod_string = odd_mod[current_number_size];
+         else
+            mod_string = even_mod[current_number_size];
+
+         standard_number a_s(a_string), b_s(b_string), mod_s(mod_string), result_s(0);
+         params_number   mod(mod_s);
+         modular_number  a(a_s, mod), b(b_s, mod), result(0, mod);
+
+         BOOST_CHECK_EQUAL((a & b).template convert_to<standard_number>(), (((a_s % mod_s) & (b_s % mod_s))) % mod_s);
+         BOOST_CHECK_EQUAL((a | b).template convert_to<standard_number>(), (((a_s % mod_s) | (b_s % mod_s))) % mod_s);
+         BOOST_CHECK_EQUAL((a ^ b).template convert_to<standard_number>(), (((a_s % mod_s) ^ (b_s % mod_s))) % mod_s);
+      }
+   }
+}
+
+template <typename Backend>
+void pow_test()
+{
+   typedef number<modular_adaptor<Backend> > modular_number;
+   typedef modular_params<Backend>           params_number;
+   typedef number<Backend>                   standard_number;
+
+   for (size_t current_number_size = 0; current_number_size < 7; ++current_number_size)
+   {
+      for (size_t parity = 0; parity <= 1; ++parity)
+      {
+         std::string a_string = big_numbers[0].substr(0, numbers_size[current_number_size]);
+         std::string b_string = big_numbers[1].substr(0, numbers_size[current_number_size]);
+         std::string mod_string;
+         if (parity)
+            mod_string = odd_mod[current_number_size];
+         else
+            mod_string = even_mod[current_number_size];
+
+         standard_number a_s(a_string), b_s(b_string), mod_s(mod_string), result_s(0);
+         params_number   mod(mod_s);
+         modular_number  a(a_s, mod), b(b_s, mod), result(0, mod);
+         BOOST_CHECK_EQUAL(pow(a, b).template convert_to<standard_number>(), powm(a_s % mod_s, b_s % mod_s, mod_s));
+      }
+   }
+}
+
+template <typename Backend>
+void run_tests()
+{
+   base_operators<Backend>();
+   comparsion_operators<Backend>();
+   bitwise_operators<Backend>();
+   pow_test<Backend>();
+}
+
+int main()
```

Comment:
Small, and rather pedantic point here: not withstanding the fact that I've written many tests like this, I've come to the conclusion that it's much more convenient to put the templates in a header and then write separate - and very short - .cpp files for each type tested.  It's just that it makes it easier to run and debug individual tests when you're not using the supplied build system (whether Boost.Build or CMake).

Reply:
Coming to tests.

I actually think most of tests should be rewritten to modern Boost.Test usage with all these ```BOOST_AUTO_TEST_SUITE``` and ```BOOST_FIXTURE_TEST_CASE```. So this particular solution you see is a really dumb one - we made it just to put some stub and make it work for now only.